### PR TITLE
Introduce binding_exprt::instantiate

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/std_expr.cpp [hunk 1]
```diff
@@ -134,3 +136,136 @@ void let_exprt::validate(const exprt &expr, const validation_modet vm)
...
+  std::map<symbol_exprt, exprt> value_map;
+
+  for(std::size_t i = 0; i < variables.size(); i++)
+  {
+    // types must match
+    PRECONDITION(variables[i].type() == values[i].type());
+    value_map[variables[i]] = values[i];
+  }
+
+  // build a subsitution map
```
Comment (peterschrammel) [inline src/util/std_expr.cpp]:
```suggestion
  // build a substitution map
```

File: src/util/std_expr.cpp [hunk 2]
```diff
@@ -134,3 +136,136 @@ void let_exprt::validate(const exprt &expr, const validation_modet vm)
       "let bindings must be type consistent");
   }
 }
+
+static optionalt<exprt> substitute_symbols_rec(
+  const std::map<irep_idt, exprt> &substitutions,
+  exprt src)
```
Comment (peterschrammel) [inline src/util/std_expr.cpp:142]: Is it necessary to pass by value here (a copy is taken below)?

**Review by peterschrammel — APPROVED** [review]

File: src/util/std_expr.cpp (hunk 2 above)
Comment (kroening) [reply src/util/std_expr.cpp:142]: The copy happens in most cases -- so this is the pattern "use pass by value when copy is inevitable".

File: src/util/std_expr.cpp (hunk 1 above)
Comment (kroening) [reply src/util/std_expr.cpp]: Fixed

**Review by TGWDB — APPROVED** [review]

**Pull request merged**